src/utils/cspUtils.js

```diff
@@ -0,0 +1,133 @@
+  if (_stringContainsValue(policy, "'none'")) return false; 
+  
+  const selfSatisfied = _checkSelfPolicy(elemAttr, policy);
+  const schemasourceSatisfied = _checkSchemasourcePolicy(elemAttr, policy);
+  const hostsourceSatisfied = _checkHostsourcePolicy(elemAttr, policy);
+  return selfSatisfied || schemasourceSatisfied || hostsourceSatisfied;
+};
+
+/* Helper function to get resource policy from csp for given policyType*/
+function _getResourcePolicy(csp, policyType) {
+  const resourcePolicyMapping = { 
+    'frame-src': ['frame', 'iframe'],
+    'img-src': ['img'],
+    'media-src': ['audio', 'video', 'track'],
+    'object-src': ['object', 'embed', 'applet'],
+    'script-src-elem': ['script'],
+  };
+  const cspPolicy = new Policy(csp);
+
+  let resourcePolicy
+  if (policyType === 'frame-src') { // from http://csplite.com/csp/test121/, fallback chain: frame-src -> child-src -> default-src
+    resourcePolicy = cspPolicy.get(policyType) ||  cspPolicy.get('child-src') || cspPolicy.get('default-src')
+  } else if (policyType === 'script-src-elem') { // fallback chain: script-src-elem -> script-src -> default-src
+    resourcePolicy = cspPolicy.get(policyType) ||  cspPolicy.get('script-src') || cspPolicy.get('default-src')
+  } else {
+    resourcePolicy = cspPolicy.get(policyType) === '' ? cspPolicy.get('default-src') : cspPolicy.get(policyType);
+  }
+  const resourcePolicyElems = resourcePolicyMapping[policyType];
+  return { resourcePolicy, resourcePolicyElems } 
+}
+
+/* Helper function to check if resourcePolicyElems for a specific policyType satisfies CSP source specifications */
+function _checkResourcePolicyElems(resourcePolicyElems, policy, $) {
+  let policyViolation = false;
+  resourcePolicyElems.forEach(elemType => { 
+    $(elemType).each((i, elem) => {
+      const checkAttr = elemType === 'object' ? 'data' : 'src' // exception for object html: <object data='abc.html'></object>
+      if (!_elemAttrSatisfiesPolicies($(elem).attr(checkAttr), policy)) {
+        $(elem).replaceWith(`<span style="color:#c91508"> Intended &lt${elemType}&gt content violates Content Security Policy and therefore could not be displayed. Isomer does not support display of any forbidden resources. </span>`);
+        policyViolation = true;
+      }
+    });
+  });
+  return { $, policyViolation };
+};
+
+export function checkCSP(
+  csp,
+  chunk,
+  policyTypes = ['img-src', 'media-src', 'frame-src', 'object-src', 'script-src-elem'],
+  ) {
+
+  let $ = cheerio.load(chunk);
+  let isCspViolation = false;
+
+  policyTypes.forEach(policyType => {
+    let policyViolation;
+    const { resourcePolicy, resourcePolicyElems } = _getResourcePolicy(csp, policyType)
+    console.log(`checking resource policy for ${policyType}, for elements ${resourcePolicyElems}: ${resourcePolicy}`);
+    ({ $, policyViolation } = _checkResourcePolicyElems(resourcePolicyElems, resourcePolicy, $));
+    isCspViolation = policyViolation || isCspViolation;
+  })
```

Comment:
You can use a reducer function for this section, with the initial value being the `$` variable loaded from the chunk https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

Reply:
e259976